Mathlib/Tactic/Tauto.lean

```diff
@@ -29,9 +29,9 @@ initialize registerTraceClass `tauto
 def distribNotOnceAt (hypFVar : Expr) (g : MVarId) : MetaM AssertAfterResult := g.withContext do
   let .fvar fvarId := hypFVar | throwError "not fvar {hypFVar}"
   let h ← fvarId.getDecl
-  let e : Q(Prop) ← (do guard (← inferType h.type).isProp; pure h.type)
+  let e : Q(Prop) ← (do guard <| ← Meta.isProp h.type; whnfR h.type)
```

Comment:
Could you say why the `whnf` is needed here?

Reply:
I figured we wanted to clean up any annotations or metadata, and reducible whnf will do at least that. I also figured that if there are abbreviations we wouldn't mind if tauto could reach into them, but I really don't have a firm grasp of how to design around whnf/cleanupAnnotations/etc. At least `Not` doesn't get unfolded by `whnfR`.

Reply:
Ok, `whnfR` was wrong, since `tauto` is used on arbitrary un-pruned goal states, with hypotheses that might not have anything to do with what `tauto` is expected to look at.

After a bit more experimentation, I discovered that Qq `match` expressions are actually doing matching up to reducible defeq (!) so there's no need to instantiate metavariables. The real problem was the check that was using `Lean.Expr.isProp` rather than the more-accurate `Lean.Meta.isProp`.